### PR TITLE
Ingester: Added configuration to set the size of the tsdb in-memory queue used before flushing chunks to the disk

### PR DESCRIPTION
**Pull request by t00350320**
**What this PR does**:
 Added configuration to set the size of the tsdb in-memory queue used before flushing chunks to the disk
**Which issue(s) this PR fixes**:
Fixes #<issue number>
https://github.com/cortexproject/cortex/issues/4957

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by t00350320** [comment]
configuration is added in cortex.yaml like this:
```
blocks_storage:
  tsdb:
    xxx
    head_chunks_write_queue_size: 5000
```
corresponding metric 'cortex_ingester_tsdb_chunk_write_queue_operations_total' is also exported.

FYI, some basic test datas ：
4C32G 200G *3 cluster,   3.6M series /30s input, replica factor = 2

For ”head_chunks_write_queue_size“ ，
1、from 0 to 1000, write latency seems better , but it 
becomes worse when from 1000 to 5000. 
2、from 0 to 1000 , to 5000, the disk busy io seems pretty much the same

For "Distributor Requests/sec"
because our prometheus remote write's queue_config
```
max_samples_per_send: 500
```
so, 3.6M series / 500 /30s =  240( Distributor Requests/sec) 

For "Write queue"
40000*60s=2.4M=3.6M * 2[replica factor] / 3[nodes number]

## case:
### head_chunks_write_queue_size = 0,
write latency:
image
cortex_ingester_tsdb_chunk_write_queue_operations_total stat:
image
disk busy:
image
### head_chunks_write_queue_size = 1000,
write latency:
image
cortex_ingester_tsdb_chunk_write_queue_operations_total stat:
image
disk busy:
image
### head_chunks_write_queue_size = 5000,
write latency:
image
cortex_ingester_tsdb_chunk_write_queue_operations_total stat:
image
disk busy:
image

**Comment by yeya24** [comment]
Nice! @t00350320 Could you please sign the DCO and fix the lint?

File: CHANGELOG.md [hunk 1]
```diff
@@ -73,6 +73,7 @@
 * [FEATURE] QueryFrontend: Support vertical sharding for subqueries. #4955
 * [FEATURE] Querier: Added a new limit `-querier.max-fetched-data-bytes-per-query` allowing to limit the maximum size of all data in bytes that a query can fetch from each ingester and storage. #4854
 * [FEATURE] Added 2 flags `-alertmanager.alertmanager-client.grpc-compression` and `-querier.store-gateway-client.grpc-compression` to configure compression methods for grpc clients. #4889
+* [FEATURE] Ingester: Added `-blocks-storage.tsdb.head-chunks-write-queue-size` allowing to configure the size of the in-memory queue used before flushing chunks to the disk . #5000
```
Comment (yeya24) [inline CHANGELOG.md:76]: The changelog entry should go to `unreleased` section.

File: pkg/storage/tsdb/config.go [hunk 2]
```diff
@@ -136,6 +136,8 @@ type TSDBConfig struct {
 	WALSegmentSizeBytes       int           `yaml:"wal_segment_size_bytes"`
 	FlushBlocksOnShutdown     bool          `yaml:"flush_blocks_on_shutdown"`
 	CloseIdleTSDBTimeout      time.Duration `yaml:"close_idle_tsdb_timeout"`
+	//The size of the in-memory queue used before flushing chunks to the disk.
+	HeadChunksWriteQueueSize int `yaml:"head_chunks_write_queue_size"`
```
Comment (yeya24) [inline pkg/storage/tsdb/config.go:140]: Please also expose this as a flag in `RegisterFlags`.

**Review by yeya24 — APPROVED** [review]
This looks good! Thanks for the contribution. Would you mind resolving the conflict again? @t00350320 Thanks!

**Comment by yeya24** [comment]
Maybe @alanprot @friedrichg would like to take another look?

File: pkg/storage/tsdb/config.go [hunk 3]
```diff
@@ -136,6 +136,8 @@ type TSDBConfig struct {
 	WALSegmentSizeBytes       int           `yaml:"wal_segment_size_bytes"`
 	FlushBlocksOnShutdown     bool          `yaml:"flush_blocks_on_shutdown"`
 	CloseIdleTSDBTimeout      time.Duration `yaml:"close_idle_tsdb_timeout"`
+	//The size of the in-memory queue used before flushing chunks to the disk.
```
Comment (yeya24) [inline pkg/storage/tsdb/config.go:139]: This needs one space between `//` and the comment.

**Comment by yeya24** [comment]
This pr LGTM. It would be appreciated if we could have another 👁️ on it. WDYT? @alanprot @friedrichg

**Review by alanprot — APPROVED** [review]

**Comment by alanprot** [comment]
Awesome! thanks!

**Pull request merged**